### PR TITLE
RM-7417: Nullable Reference Types in Core.UnitTests

### PR DESCRIPTION
**Pull request by dkudernatsch**

### CODE REVIEW

File: Remotion/Core/UnitTests/DoubleCheckedLockingContainerTest.cs [hunk 1]
```diff
@@ -78,8 +78,8 @@ public void SetNull()
     {
       SampleClass expected = new SampleClass ();
       IFactory mockFactory = _mocks.StrictMock<IFactory> ();
-      DoubleCheckedLockingContainer<SampleClass> container =
-          new DoubleCheckedLockingContainer<SampleClass> (delegate { return mockFactory.Create (); });
+      DoubleCheckedLockingContainer<SampleClass?> container =
```
Comment (MichaelKetting) [inline Remotion/Core/UnitTests/DoubleCheckedLockingContainerTest.cs:81]: update after Core.Core

File: Remotion/Core/UnitTests/Reflection/TestDomain/PropertyInfoExtensions/DerivedOpenGenericClassWithDifferentProperties.cs [hunk 2]
```diff
@@ -15,14 +15,15 @@
 // along with re-motion; if not, see http://www.gnu.org/licenses.
 // 
 using System;
-
+using System.Diagnostics.CodeAnalysis;
+#nullable disable
```
Comment (MichaelKetting) [inline Remotion/Core/UnitTests/Reflection/TestDomain/PropertyInfoExtensions/DerivedOpenGenericClassWithDifferentProperties.cs:19]: why?

File: Remotion/Core/UnitTests/Reflection/TestDomain/PropertyInfoExtensions/GenericClassWithDifferentProperties.cs [hunk 3]
```diff
@@ -15,6 +15,9 @@
 // along with re-motion; if not, see http://www.gnu.org/licenses.
 // 
 using System;
+using System.Diagnostics.CodeAnalysis;
+
+#nullable disable
```
Comment (MichaelKetting) [inline Remotion/Core/UnitTests/Reflection/TestDomain/PropertyInfoExtensions/GenericClassWithDifferentProperties.cs:20]: Why?

File: Remotion/Core/UnitTests/ServiceLocation/SafeServiceLocatorTest.cs [hunk 4]
```diff
@@ -237,7 +237,7 @@ private void ResetDefaultServiceLocator ()
     {
       var defaultServiceLocatorContainer = 
           (DoubleCheckedLockingContainer<IServiceLocator>) PrivateInvoke.GetNonPublicStaticField (typeof (SafeServiceLocator), "s_defaultServiceLocator");
-      defaultServiceLocatorContainer.Value = null;
+      defaultServiceLocatorContainer.Value = default!;
```
Comment (MichaelKetting) [inline Remotion/Core/UnitTests/ServiceLocation/SafeServiceLocatorTest.cs:240]: revert

File: Remotion/Core/UnitTests/Utilities/EnumTypeConverterFactoryTest.cs [hunk 5]
```diff
@@ -42,16 +42,16 @@ public void SetUp ()
     public void CreateTypeConverterOrDefault_WithEnum_ReturnsAdvancedEnumConverter ()
     {
       var typeConverter = _factory.CreateTypeConverterOrDefault (typeof (Int32Enum));
-      Assert.That (typeConverter, Is.TypeOf<AdvancedEnumConverter>());
-      Assert.That (((AdvancedEnumConverter) typeConverter).EnumType, Is.EqualTo (typeof (Int32Enum)));
+      Assert.That (typeConverter, Is.TypeOf<AdvancedEnumConverter?>());
```
Comment (MichaelKetting) [inline Remotion/Core/UnitTests/Utilities/EnumTypeConverterFactoryTest.cs:45]: why?

File: Remotion/Core/UnitTests/Utilities/EnumTypeConverterFactoryTest.cs [hunk 6]
```diff
@@ -42,16 +42,16 @@ public void SetUp ()
     public void CreateTypeConverterOrDefault_WithEnum_ReturnsAdvancedEnumConverter ()
     {
       var typeConverter = _factory.CreateTypeConverterOrDefault (typeof (Int32Enum));
```
Comment (MichaelKetting) [inline Remotion/Core/UnitTests/Utilities/EnumTypeConverterFactoryTest.cs:44]: NFO, then we shouldn't need the null-work in line 46

File: Remotion/Core/UnitTests/Utilities/EnumTypeConverterFactoryTest.cs [hunk 7]
```diff
@@ -42,16 +42,16 @@ public void SetUp ()
...
-      Assert.That (typeConverter, Is.TypeOf<AdvancedEnumConverter>());
-      Assert.That (((AdvancedEnumConverter) typeConverter).EnumType, Is.EqualTo (typeof (Int32Enum)));
+      Assert.That (typeConverter, Is.TypeOf<AdvancedEnumConverter?>());
+      Assert.That (((AdvancedEnumConverter?) typeConverter)!.EnumType, Is.EqualTo (typeof (Int32Enum)));
     }
 
     [Test]
     public void CreateTypeConverterOrDefault_WithNullableEnum_ReturnsAdvancedEnumConverter ()
     {
       var typeConverter = _factory.CreateTypeConverterOrDefault (typeof (Int32Enum?));
```
Comment (MichaelKetting) [inline Remotion/Core/UnitTests/Utilities/EnumTypeConverterFactoryTest.cs:52]: NFO

File: Remotion/Core/UnitTests/Utilities/NameValueCollectionUtilityTest.cs [hunk 8]
```diff
@@ -136,7 +136,7 @@ public void MergeWithFirstNullAndSecondNull()
   [Test]
   public void MergeWithFirstValueAndSecondNull()
   {
-    NameValueCollection actual = NameValueCollectionUtility.Merge (_collection, null);
+    NameValueCollection actual = NameValueCollectionUtility.Merge (_collection, null)!;
```
Comment (MichaelKetting) [inline Remotion/Core/UnitTests/Utilities/NameValueCollectionUtilityTest.cs:139]: revert here and below after updating core.core annotations for Merge

File: Remotion/Core/UnitTests/Utilities/TypeUtilityTests.cs [hunk 9]
```diff
@@ -357,7 +357,7 @@ public void GetAbbreviatedTypeName_WithClosedNestedGenericNestedGenericType ()
 
     private void AssertTransformation(string abbreviatedName, string fullName)
     {
-      string result = TypeUtility.ParseAbbreviatedTypeName (abbreviatedName);
+      string? result = TypeUtility.ParseAbbreviatedTypeName (abbreviatedName);
```
Comment (MichaelKetting) [inline Remotion/Core/UnitTests/Utilities/TypeUtilityTests.cs:360]: is annotated, should not be needed, right?

File: Remotion/Core/UnitTests/Utilities/TypeUtilityTests.cs [hunk 10]
```diff
@@ -121,21 +121,21 @@ public void TestAbbreviatedClosedGenericTypeWithPartiallyAbbreviatedTypeArgument
     [Test]
     public void TestGetType ()
     {
-      Type t = TypeUtility.GetType ("Remotion.UnitTests::Utilities.TypeUtilityTests", true);
+      Type t = TypeUtility.GetType ("Remotion.UnitTests::Utilities.TypeUtilityTests", true)!;
```
Comment (MichaelKetting) [inline Remotion/Core/UnitTests/Utilities/TypeUtilityTests.cs:124]: should not be needed after Core.Core annotations are done. here and below.

File: Remotion/Core/UnitTests/Collections/ReadOnlyCollectionDecoratorTest.cs [hunk 11]
```diff
@@ -24,12 +24,12 @@ namespace Remotion.UnitTests.Collections
   [TestFixture]
   public class ReadOnlyCollectionDecoratorTest
   {
-    private ReadOnlyCollectionDecorator<string >_collection;
+    private ReadOnlyCollectionDecorator<string?>_collection = default!;
```
[the remark this answers is not in the document]
Comment (dkudernatsch) [reply Remotion/Core/UnitTests/Collections/ReadOnlyCollectionDecoratorTest.cs:27]: `_collection.Contains (null)` in a test below requires this to be nullable

**Comment by iam-mholle** [comment]
```text
  SafeServiceLocatorTest.cs(38, 90): [CS8625] Cannot convert null literal to non-nullable reference type.
  SafeServiceLocatorTest.cs(51, 48): [CS8625] Cannot convert null literal to non-nullable reference type.
  SafeServiceLocatorTest.cs(59, 48): [CS8604] Possible null reference argument for parameter 'configuration' in 'void ServiceLocationConfiguration.SetCurrent(IServiceLocationConfiguration configuration)'.
  SafeServiceLocatorTest.cs(200, 20): [CS8625] Cannot convert null literal to non-nullable reference type.
  CompoundValueEqualityComparerTest.cs(63, 48): [CS8625] Cannot convert null literal to non-nullable reference type.
  ComponentwiseEqualsAndHashcodeWrapperTest.cs(65, 67): [CS8625] Cannot convert null literal to non-nullable reference type.
  DictionaryExtensionsTest.cs(88, 24): [CS8714] The type 'string?' cannot be used as type parameter 'TKey' in the generic type or method 'DictionaryExtensions.GetValueOrDefault<TKey, TValue>(IReadOnlyDictionary<TKey, TValue>, TKey)'. Nullability of type argument 'string?' doesn't match 'notnull' constraint.
  DictionaryExtensionsTest.cs(98, 24): [CS8714] The type 'string?' cannot be used as type parameter 'TKey' in the generic type or method 'DictionaryExtensions.GetValueOrDefault<TKey, TValue>(IDictionary<TKey, TValue>, TKey)'. Nullability of type argument 'string?' doesn't match 'notnull' constraint.
  AssemblyFinderTypeDiscoveryServiceTest.cs(108, 45): [CS8625] Cannot convert null literal to non-nullable reference type.
  AssemblyFinderTypeDiscoveryServiceTest.cs(156, 45): [CS8625] Cannot convert null literal to non-nullable reference type.
  DictionaryExtensionsTest.cs(193, 24): [CS8714] The type 'string?' cannot be used as type parameter 'TKey' in the generic type or method 'DictionaryExtensions.GetOrCreateValue<TKey, TValue>(IDictionary<TKey, TValue>, TKey, Func<TKey, TValue>)'. Nullability of type argument 'string?' doesn't match 'notnull' constraint.
  DelegateBasedEqualityComparerTest.cs(53, 45): [CS8625] Cannot convert null literal to non-nullable reference type.
  DelegateBasedEqualityComparerTest.cs(59, 38): [CS8625] Cannot convert null literal to non-nullable reference type.
  DelegateBasedEqualityComparerTest.cs(59, 44): [CS8625] Cannot convert null literal to non-nullable reference type.
  StubProviderHelper.cs(56, 14): [CS8603] Possible null reference return.
  StubProviderHelper.cs(61, 14): [CS8603] Possible null reference return.
  StubProviderHelper.cs(87, 71): [CS8604] Possible null reference argument for parameter 'defaultValue' in 'ConfigurationProperty ProviderHelperBase<IFakeProvider>.CreateDefaultProviderNameProperty(string name, string defaultValue)'.
  SafeContextSingletonTest.cs(92, 29): [CS8625] Cannot convert null literal to non-nullable reference type.
  Log4NetTraceListenerTest.cs(149, 29): [CS8625] Cannot convert null literal to non-nullable reference type.
  Log4NetTraceListenerTest.cs(160, 29): [CS8625] Cannot convert null literal to non-nullable reference type.
  Log4NetTraceListenerTest.cs(171, 29): [CS8625] Cannot convert null literal to non-nullable reference type.
  Log4NetTraceListenerTest.cs(215, 28): [CS8625] Cannot convert null literal to non-nullable reference type.
  Log4NetTraceListenerTest.cs(230, 28): [CS8625] Cannot convert null literal to non-nullable reference type.
  Log4NetTraceListenerTest.cs(280, 32): [CS8625] Cannot convert null literal to non-nullable reference type.
  Log4NetTraceListenerTest.cs(295, 32): [CS8625] Cannot convert null literal to non-nullable reference type.
  Log4NetTraceListenerTest.cs(295, 49): [CS8625] Cannot convert null literal to non-nullable reference type.
```

File: Remotion/Core/UnitTests/Collections/DictionaryExtensionsTest.cs [hunk 12]
```diff
@@ -82,7 +82,7 @@ public void GetValueOrDefault_WithReadOnlyDictionary_ValueTypes ()
     [Test]
     public void GetValueOrDefault_WithIReadOnlyDictionary_NullKey ()
     {
-      var dictionaryStub = MockRepository.GenerateStub<IReadOnlyDictionary<string, string>>();
+      var dictionaryStub = MockRepository.GenerateStub<IReadOnlyDictionary<string?, string>>();
```
[the remark this answers is not in the document]
Comment (iam-mholle) [reply Remotion/Core/UnitTests/Collections/DictionaryExtensionsTest.cs:85]: I guess this is obsolete now, as nullable keys are a thing apparently.

[the remark this answers is not in the document]
Comment (MichaelKetting) [reply Remotion/Core/UnitTests/Collections/DictionaryExtensionsTest.cs:85]: correct

File: Remotion/Core/UnitTests/Utilities/TypeUtilityTests.cs (hunk 10 above)
Comment (iam-mholle) [reply Remotion/Core/UnitTests/Utilities/TypeUtilityTests.cs:124]: Unfortunately those are still needed, as there is currently no way to link a `bool` parameter with the nullability of the return type.

File: Remotion/Core/UnitTests/Context/SafeContextSingletonTest.cs [hunk 13]
```diff
@@ -89,7 +89,7 @@ public void HasCurrent ()
       Assert.That (singleton.HasCurrent, Is.False);
       Dev.Null = singleton.Current;
       Assert.That (singleton.HasCurrent, Is.True);
-      singleton.SetCurrent (null);
+      singleton.SetCurrent (null!);
```
Comment (MichaelKetting) [inline Remotion/Core/UnitTests/Context/SafeContextSingletonTest.cs:92]: shoudnt setcurrent be nullable?

File: Remotion/Core/UnitTests/ServiceLocation/SafeServiceLocatorTest.cs [hunk 14]
```diff
@@ -48,15 +48,15 @@ public void TestFixtureTearDown ()
     public void SetUp ()
     {
       _previousConfiguration = ServiceLocationConfiguration.Current;
-      ServiceLocationConfiguration.SetCurrent (null);
+      ServiceLocationConfiguration.SetCurrent (null!);
```
Comment (MichaelKetting) [inline Remotion/Core/UnitTests/ServiceLocation/SafeServiceLocatorTest.cs:51]: SetCurrent should be nullable.

**Pull request merged**